airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobCleaner.java

```diff
@@ -0,0 +1,165 @@
+ * this two ways. 1. It cleans out all files and directories that are older than the maximum
+ * retention date. 2. It cleans out the oldest files before the minimum retention date until it is
+ * within the max workspace size.
+ */
+public class JobCleaner implements Runnable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(JobCleaner.class);
+
+  private final Path workspaceRoot;
+  private final JobPersistence jobPersistence;
+
+  private final long minWorkspaceRetentionDays;
+  private final long maxWorkspaceRetentionDays;
+  private final long maxWorkspaceSizeMb;
+
+  public JobCleaner(long minWorkspaceRetentionDays,
+                    long maxWorkspaceRetentionDays,
+                    long maxWorkspaceSizeMb,
+                    Path workspaceRoot,
+                    JobPersistence jobPersistence) {
+    this.minWorkspaceRetentionDays = minWorkspaceRetentionDays;
+    this.maxWorkspaceRetentionDays = maxWorkspaceRetentionDays;
+    this.maxWorkspaceSizeMb = maxWorkspaceSizeMb;
+    this.workspaceRoot = workspaceRoot;
+    this.jobPersistence = jobPersistence;
+  }
+
+  @Override
+  public void run() {
+    try {
+      deleteOldFiles();
+      deleteOnSize();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void deleteOldFiles() throws IOException {
+    final Date oldestAllowed = getDateFromDaysAgo(maxWorkspaceRetentionDays);
+
+    Files.walk(workspaceRoot)
+        .map(Path::toFile)
+        .filter(f -> new AgeFileFilter(oldestAllowed).accept(f))
+        .forEach(file -> {
+          LOGGER.info("Deleting old file: " + file.toString());
+          FileUtils.deleteQuietly(file);
+
+          File parentDir = file.getParentFile();
+          if (parentDir.isDirectory() && parentDir.listFiles().length == 0) {
+            FileUtils.deleteQuietly(parentDir);
+          }
+        });
+  }
+
+  private void deleteOnSize() throws IOException {
+    Set<String> nonTerminalJobIds = new HashSet<>();
+    Sets.SetView<JobStatus> nonTerminalStatuses = Sets.difference(Set.of(JobStatus.values()), JobStatus.TERMINAL_STATUSES);
+
+    for (JobStatus nonTerminalStatus : nonTerminalStatuses) {
+      Set<String> jobIds = jobPersistence.listJobsWithStatus(nonTerminalStatus)
+          .stream()
+          .map(Job::getId)
+          .map(String::valueOf)
+          .collect(Collectors.toSet());
+
+      nonTerminalJobIds.addAll(jobIds);
+    }
+
+    final Date youngestAllowed = getDateFromDaysAgo(minWorkspaceRetentionDays);
+
+    final long workspaceBytes = FileUtils.sizeOfDirectory(workspaceRoot.toFile());
+    AtomicLong deletedBytes = new AtomicLong(0);
+    final AgeFileFilter ageFilter = new AgeFileFilter(youngestAllowed);
+    Files.walk(workspaceRoot)
+        .map(Path::toFile)
+        .filter(f -> {
+          String[] splits = f.toPath().toString().replace(workspaceRoot.toString(), "").split("/");
```

Comment:
any way to avoid splitting on `/` since it would break on windows?

Reply:
For all the past manipulation we should use the provided path api that will take care of all these details for us